src/merge.ts

```diff
@@ -1,24 +1,114 @@
-import { IHttpHeaderParam, IHttpOperation, IMediaTypeContent, IServer } from '@stoplight/types';
+import {
+  IHttpOperation,
+  IHttpOperationRequestBody,
+  IHttpParam,
+  IMediaTypeContent,
+  INodeExample,
+  INodeExternalExample,
+  IServer,
+} from '@stoplight/types';
+import { JSONSchema4, JSONSchema6, JSONSchema7 } from 'json-schema';
+import { isEqual } from 'lodash';
 
-const mergeHeaders = mergeLists<IHttpHeaderParam[]>(
-  (h1, h2) => h1.name.toLowerCase() === h2.name.toLowerCase(),
-  h1 => h1, // ignore header #2 if mediaTypes is equal
-);
+type JSONSchema = JSONSchema4 | JSONSchema6 | JSONSchema7;
+
+function isSimpleAnyOfSchema(schema: JSONSchema) {
+  return schema.anyOf && Object.keys(schema).length === 1;
+}
+
+function mergeSchemas(schema1: JSONSchema, schema2: JSONSchema): JSONSchema {
+  const schemas = (isSimpleAnyOfSchema(schema2) ? (schema2.anyOf as JSONSchema[]) : [schema2]).reduce<
+    Array<JSONSchema>
+  >(
+    (schemas, schema) => {
+      if (!schemas.find(s => isEqual(s, schema))) {
+        schemas.push(schema);
+      }
+      return schemas;
+    },
+    isSimpleAnyOfSchema(schema1) ? (schema1.anyOf as JSONSchema[]) : [schema1],
+  );
+
+  return schemas.length === 1 ? schemas[0] : { anyOf: schemas as any };
+}
+
+function mergeParams<T extends IHttpParam>(params1: T[], params2: T[]): T[] {
+  const params1OnlyAndCommon = params1.map(p1 => {
+    const p2 = params2.find(p2 => p2.name.toLowerCase() === p1.name.toLowerCase());
+
+    if (p2) {
+      return {
+        ...p2,
+        ...p1,
+        required: true,
+        schema: p1.schema && p2.schema ? mergeSchemas(p1.schema, p2.schema) : p1.schema ? p1.schema : p2.schema,
+      };
+    } else {
+      return {
+        ...p1,
+        required: false,
+      };
+    }
+  });
+
+  const params2Only = params2
+    .filter(p2 => !params1.find(p1 => p1.name.toLowerCase() === p2.name.toLowerCase()))
+    .map(h2 => ({ ...h2, required: false }));
```

Comment:
Why are these parameters not required?

Reply:
Because it does not exists in `params1` list.

Params1:
```
Header1: ..
Header2: ..
```

Params2:
```
Header2: ..
Header3: ..
```

`params1OnlyAndCommon = [Header1,Header2]`
`params2Only = [Header3]`

Required: `[Header2]`
NonRequired: `[Header1,Header3]`

The `params1OnlyAndCommon` thingy is an optimization to cover those two cases with a single loop.

Reply:
Ah ok understood. So essentially the strategy for the merging is that if a parameter is present on *both* headers then it's required, otherwise it's not, correct?

Reply:
Yes, more here: https://github.com/stoplightio/http-spec/pull/110#discussion_r462264838